test/jdk/java/util/concurrent/tck/ForkJoinPool20Test.java

```diff
@@ -216,4 +238,416 @@ static ForkJoinWorkerThread submitBusyTask(ForkJoinPool pool) throws Exception {
+            tasks.add(p.scheduleWithFixedDelay(r, 9, 9, SECONDS));
+        }
+        p.shutdownNow();
+        assertTrue(p.awaitTermination(LONG_DELAY_MS, MILLISECONDS));
+        for (ScheduledFuture<?> task : tasks) {
+            assertTrue(task.isDone());
+        }
+        assertTrue(p.isTerminated());
+    }
+
+    /**
+     * submitWithTimeout cancels task after timeout
+     */
+    public void testSubmitWithTimeoutCancels() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return Boolean.TRUE; }};
+        ForkJoinTask<?> task = p.submitWithTimeout(c, 1, NANOSECONDS, null);
+        Thread.sleep(timeoutMillis());
+        assertTrue(task.isCancelled());
+    }
+
+    static final class SubmitWithTimeoutException extends RuntimeException {}
+
+    /**
+     * submitWithTimeout using complete completes after timeout
+     */
+    public void testSubmitWithCompleterTimeoutCompletes() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Item> c = new Callable<Item>() {
+            public Item call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return one; }};
+        ForkJoinTask<Item> task = p.submitWithTimeout(
+            c, 1, NANOSECONDS,
+            (ForkJoinTask<Item> t) ->
+            t.complete(two));
+        Thread.sleep(timeoutMillis());
+        assertEquals(task.join(), two);
+    }
+
+    /**
+     * submitWithTimeout using completeExceptionally throws after timeout
+     */
+    public void testSubmitWithTimeoutThrows() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return Boolean.TRUE; }};
+        ForkJoinTask<Boolean> task = p.submitWithTimeout(
+            c, 1, NANOSECONDS,
+            (ForkJoinTask<Boolean> t) ->
+            t.completeExceptionally(new SubmitWithTimeoutException()));
+        Thread.sleep(timeoutMillis());
+        try {
+            task.join();
+            shouldThrow();
+        }
+        catch (Exception ex) {
+            assertTrue(ex instanceof SubmitWithTimeoutException);
+        }
+    }
+
+    /**
+     * submitWithTimeout doesn't cancel if completed before timeout
+     */
+    public void testSubmitWithTimeout_NoTimeout() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                return Boolean.TRUE; }};
+        ForkJoinTask<?> task = p.submitWithTimeout(c, LONGER_DELAY_MS, MILLISECONDS, null);
+        Thread.sleep(timeoutMillis());
```

Comment:
Suggest to check that the task was successfully completed with the right result (as well).

Reply:
done.